src/main/java/com/netgrif/application/engine/petrinet/service/UriService.java

```diff
@@ -0,0 +1,239 @@
+     * Moves UriNode to other destination
+     *
+     * @param uri     to be moved
+     * @param destUri the destination URI
+     * @return result UriNode object
+     */
+    @Override
+    public UriNode move(String uri, String destUri) {
+        UriNode uriNode = findByUri(uri);
+        return move(uriNode, destUri);
+    }
+
+    /**
+     * Moves UriNode to other destination
+     *
+     * @param node    to be moved
+     * @param destUri the destination URI
+     * @return result UriNode object
+     */
+    @Override
+    public UriNode move(UriNode node, String destUri) {
+        UriNode newParent = getOrCreate(destUri, null);
+        UriNode oldParent = findById(node.getParentId());
+
+        oldParent.getChildrenId().remove(node.getId());
+        newParent.getChildrenId().add(node.getId());
+
+        node.setParentId(newParent.getId());
+        node.setUriPath(destUri + uriSeparator + node.getName());
+        return uriNodeRepository.save(node);
+    }
+
+    /**
+     * Creates new UriNode from PetriNet identifier, or retrieves existing one
+     *
+     * @param petriNet    to be used for creating UriNode
+     * @param contentType to decide the content type of UriNode
+     * @return the UriNode that was created or modified
+     */
+    @Override
+    public UriNode getOrCreate(PetriNet petriNet, UriContentType contentType) {
+        String identifier = petriNet.getIdentifier();
+        String modifiedUri;
+        if (identifier.contains(uriSeparator))
+            modifiedUri = identifier.substring(0, identifier.lastIndexOf(uriSeparator));
+        else
+            modifiedUri = DEFAULT_ROOT_URI;
+
+        return getOrCreate(modifiedUri, contentType);
+    }
+
+    /**
+     * Creates new UriNode from URI path, or retrieves existing one
+     *
+     * @param uri         to be used for creating UriNode
+     * @param contentType to decide the content type of UriNode
+     * @return the UriNode that was created or modified netgrif/process/test/all_data, netgrif/process
+     */
+    @Override
+    public UriNode getOrCreate(String uri, UriContentType contentType) {
+        LinkedList<UriNode> uriNodeList = new LinkedList<>();
+        String[] uriComponents = uri.split(uriSeparator);
+        StringBuilder uriBuilder = new StringBuilder();
+        int pathLength = uriComponents.length;
+        UriNode parent = pathLength > 1 || !uri.equals(DEFAULT_ROOT_URI) ? uriNodeRepository.findByUriPath(DEFAULT_ROOT_URI) : null;
+
+        for (int i = 0; i < pathLength; i++) {
+            uriBuilder.append(uriComponents[i]);
+            UriNode uriNode = uriNodeRepository.findByUriPath(uriBuilder.toString());
+            if (uriNode == null) {
+                uriNode = new UriNode();
+                uriNode.setName(uriComponents[i]);
+                uriNode.setLevel(i + 1);
+                uriNode.setUriPath(uriBuilder.toString());
+                uriNode.setParentId(parent != null ? parent.getId() : null);
```

Comment:
if only one root node can exist at a time and this node is represented by an object instance, why is it allowed to have nodes with parents that are null?